routers/web/repo/release.go

```diff
@@ -344,59 +371,42 @@ func NewRelease(ctx *context.Context) {
 
 		if rel != nil {
 			rel.Repo = ctx.Repo.Repository
-			if err := rel.LoadAttributes(ctx); err != nil {
+			if err = rel.LoadAttributes(ctx); err != nil {
 				ctx.ServerError("LoadAttributes", err)
 				return
 			}
 
+			ctx.Data["TagNameReleaseExists"] = true
 			ctx.Data["tag_name"] = rel.TagName
-			if rel.Target != "" {
-				ctx.Data["tag_target"] = rel.Target
-			}
+			ctx.Data["tag_target"] = rel.Target
 			ctx.Data["title"] = rel.Title
 			ctx.Data["content"] = rel.Note
 			ctx.Data["attachments"] = rel.Attachments
 		}
 	}
-	ctx.Data["IsAttachmentEnabled"] = setting.Attachment.Enabled
-	assigneeUsers, err := repo_model.GetRepoAssignees(ctx, ctx.Repo.Repository)
-	if err != nil {
-		ctx.ServerError("GetRepoAssignees", err)
-		return
-	}
-	ctx.Data["Assignees"] = shared_user.MakeSelfOnTop(ctx.Doer, assigneeUsers)
-
-	upload.AddUploadContext(ctx, "release")
-
-	// For New Release page
-	PrepareBranchList(ctx)
-	if ctx.Written() {
-		return
-	}
-
-	tags, err := repo_model.GetTagNamesByRepoID(ctx, ctx.Repo.Repository.ID)
-	if err != nil {
-		ctx.ServerError("GetTagNamesByRepoID", err)
-		return
-	}
-	ctx.Data["Tags"] = tags
 
 	ctx.HTML(http.StatusOK, tplReleaseNew)
 }
 
 // NewReleasePost response for creating a release
 func NewReleasePost(ctx *context.Context) {
+	newReleaseCommon(ctx)
+	if ctx.Written() {
+		return
+	}
+
 	form := web.GetForm(ctx).(*forms.NewReleaseForm)
-	ctx.Data["Title"] = ctx.Tr("repo.release.new_release")
-	ctx.Data["PageIsReleaseList"] = true
 
-	tags, err := repo_model.GetTagNamesByRepoID(ctx, ctx.Repo.Repository.ID)
-	if err != nil {
-		ctx.ServerError("GetTagNamesByRepoID", err)
+	// first, check whether the release exists,
+	// it should be done before the form error check, because the tmpl needs "TagNameReleaseExists" to show/hide the "tag only" button
+	rel, err := repo_model.GetRelease(ctx, ctx.Repo.Repository.ID, form.TagName)
+	if err != nil && !repo_model.IsErrReleaseNotExist(err) {
+		ctx.ServerError("GetRelease", err)
 		return
 	}
-	ctx.Data["Tags"] = tags
+	ctx.Data["TagNameReleaseExists"] = rel != nil
```

Comment:
Big fan of the test refactor. Everything looks good to me, the only problem that I see now is the "Create Tag Only" problem we discussed earlier:

<img width="1296" alt="Screenshot 2025-01-13 at 10 43 07 AM" src="https://github.com/user-attachments/assets/20b3f67c-8943-479a-9d6a-42ae29469b69" />


To reproduce:
- Have an existing tag e.g. "v1.0"
- Try creating a tag with the same name


It looks like this happens because the "error form" page we return now has a `tag_name` that points to an existing `Release`, which would set the highlighted line to `true`.

Would it make sense to set this to something like:
```suggestion
			ctx.Data["TagNameReleaseExists"] = rel != nil && !rel.IsTag
```

Reply:
Hmm, I think the problem is: we shouldn't stop users from creating new "tag-only" if the error occurs, no matter whether the release exists. So I think this line should be completely removed.

Reply:
Made some changes in 57aef6ddf1c0eff7a891ac8eb470a04786c00d36

```go

	// we should still show the "tag only" button if the user clicks it, no matter the release exists or not.
	// because if error occurs, end users need to have the chance to edit the name and submit the form with "tag-only" again.
	ctx.Data["ShowCreateTagOnlyButton"] = form.TagOnly || rel == nil
```

Does it sound right?

Reply:
Yes it sounds about right

> Hmm, I think the problem is: we shouldn't stop users from creating new "tag-only" if the error occurs, no matter whether the release exists. So I think this line should be completely removed.

Hmm it looks like the only case where we do hide it is when a `tag_name` field is passed which refers to an existing tag (I guess we then hide it to avoid any confusion by users). I think it is fine to follow your alternative approach and have the button always appear since the user would be shown an error in case they do press it, but it's up to you